Userland/useradd.cpp

```diff
@@ -0,0 +1,153 @@
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <AK/String.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/CArgsParser.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+static constexpr uid_t BASE_UID = 1000;
+static constexpr gid_t USERS_GID = 100;
+const char* DEFAULT_SHELL = "/bin/sh";
+
+int main(int argc, char** argv)
+{
+    CArgsParser args_parser("useradd");
+    args_parser.add_arg("d", "home_dir", "home directory for the new user");
+    args_parser.add_arg("u", "uid", "user ID (uid) for the new user");
+    args_parser.add_arg("g", "group", "group ID (gid) for the new user");
+    args_parser.add_arg("m", "create home directory if it does not exist");
+    args_parser.add_arg("s", "shell", "name of default shell binary for the new user");
+    args_parser.add_arg("n", "name", "gecos name of the new user");
+    CArgsParserResult args = args_parser.parse(argc, argv);
+
+    if (argc <= 1) {
+        args_parser.print_usage();
+        return 0;
+    }
+
+    const char* username = argv[argc - 1]; // Always the user name!
+    uid_t uid = BASE_UID;
+    gid_t gid = USERS_GID;
+
+    // First, let's sort out the uid for the user
+    if (args.is_present("u")) {
+        bool ok;
+        uid = static_cast<uid_t>(args.get("u").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid uid %s\n", args.get("u").characters());
+            return 3;
+        }
+
+        if (getpwuid(uid)) {
+            fprintf(stderr, "uid %u already exists!\n", uid);
+            return 4;
+        }
+
+    } else {
+        for (uid = BASE_UID; getpwuid(uid); uid++) {
+        }
+    }
+
+    // Now let's sort out the gid!
+    if (args.is_present("g")) {
+        bool ok;
+        gid = static_cast<gid_t>(args.get("g").to_uint(ok));
+
+        if (!ok) {
+            fprintf(stderr, "Invalid gid %s\n", args.get("g").characters());
+            return 1;
+        }
+    }
+
+    String shell = args.is_present("s") ? args.get("s") : String(DEFAULT_SHELL);
+    String home = args.is_present("d") ? args.get("d") : String("/");
+    String gecos = args.is_present("n") ? args.get("n") : String("Anonymous");
+
+    // Let's run a quick sanity check on username
+    if (strpbrk(username, "\\/0123456789!@#$%^&*()~+=`:\n")) {
+        fprintf(stderr, "invalid character in username, %s\n", username);
+        return 1;
+    }
+
+    // Disallow names starting with _ and -
+    if (!(strpbrk(username, "_-") - username)) {
+        fprintf(stderr, "invalid username, %s\n", username);
+        return 1;
+    }
```

Comment:
* Move these sanity checks next to `const char* username = argv[argc - 1];`, before parsing other args
* You're still disallowing names such as `foo96`, they should be OK (but not `96foo`!)
* That's one weird way to check the first symbol...

Reply:
> That's one weird way to check the first symbol...

How would you recommend doing this then??

> You're still disallowing names such as foo96, they should be OK (but not 96foo!)

Can I do this with `strpbrk()`??

Reply:
> How would you recommend doing this then??

I'd go with just `username[0] == '_' || username[0] == '_'`, or (weirder, but apparently shorter) `strpbrk(username, "_-") == username` if you really want to use `strpbrk()`.

Honestly, I would just check that `isalpha(username[0])` and that `isalnum(ch) || ch == '-' || ch == '_'` for the rest of the characters.